specifyweb/frontend/js_src/lib/components/WorkBench/WbActions.tsx

```diff
@@ -15,7 +15,186 @@ import type { WbCellCounts } from './CellMeta';
 import { RollbackConfirmation } from './Components';
 import { CreateRecordSetButton } from './RecordSet';
 import { WbStatus as WbStatusComponent } from './Status';
-import type { WbStatus, WbView } from './WbView';
+import type { DialogHandlers, WbStatus, WbView } from './WbView';
+import type { Dataset } from '../WbPlanView/Wrapped';
+import type { WbMapping } from './mapping';
+import type { RA } from '../../utils/types';
+
+/* eslint-disable functional/no-this-expression */
+export class WbActionsReact {
+  // eslint-disable-next-line functional/prefer-readonly-type
+  public hasUnSavedChanges: boolean = false;
+
+  public constructor(
+    private readonly data: RA<RA<string | null>>,
+    private readonly dataset: Dataset,
+    private readonly mappings: WbMapping,
+    private readonly noUploadPlan: DialogHandlers,
+    private readonly uploadDialog: DialogHandlers,
+    private readonly statusComponent: DialogHandlers,
+    private readonly checkDeletedFail: (statusCode: number) => boolean,
+    private readonly triggerRefresh: () => void,
+    private readonly saveProgressBar: DialogHandlers,
+    private refreshInitiatedBy: React.MutableRefObject<WbStatus | undefined>,
+    private refreshInitiatorAborted: React.MutableRefObject<boolean>,
+    private readonly operationCompleted: DialogHandlers,
+    private readonly operationAborted: DialogHandlers
+  ) {}
```

Comment:
This constructor is starting to look crazy and is only going to get worse. Would appreciate some suggestions for improvement.

Reply:
This class might not even exist at the end, so not a big worry.
Otherwise, we try to avoid having more than 3-4 arguments in a function as then you could easily end up with a bug where the order of arguments was swapped accidentally, or something was missed.
Instead, if you need a lot of arguments, we make the function/constructor receive an object as it's only argument, and then that object would have the keys and values for each parameter you want to have

you could look at any react component in our codebase for an example of how to type that in typescript - https://github.com/specify/specify7/blob/59a7c1ef8ae494dcf3ca2cb409a1b8773ce5fe73/specifyweb/frontend/js_src/lib/components/WbImport/index.tsx#L205-L215